---
title: "U2OS BRCA1 RNAPII ChIP"
author: "Aiola Stoja"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: lumen
    code_folding: hide
    toc: yes
    toc_float:
      collapsed: yes
      smooth_scroll: yes
  pdf_document:
    toc: yes
---

```{r Packages, message=FALSE, warning=FALSE, include=FALSE}
library(BiocManager)
#devtools::install_github("YuLab-SMU/ChIPseeker")
library(ChIPseeker)
#BiocManager::install("TxDb.Hsapiens.UCSC.hg38.knownGene")
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
#BiocManager::install("clusterProfiler")
library(clusterProfiler)
#library(remotes) ##used to install seqUtils package to use loadNarrowPeaks (below)
#remotes::install_github("kauralasoo/seqUtils")
library(seqUtils)
library(ggpubr)
library(ggplot2)
library(ggimage)
library(ggVennDiagram)
#BiocManager::install("ReactomePA")
library(ReactomePA)
library(dplyr)
library(stringr)
#BiocManager::install("org.Hs.eg.db")
library(org.Hs.eg.db)
library(patchwork)
#install.packages("DT")
library(DT)
library(kableExtra)
library(rgl)
library(lintr)
library(VennDiagram)

```

```{r include=FALSE, setup,}
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::knit_hooks$set(webgl = hook_webgl)
```

# ChIP peak coverage

```{r ChIP peaks coverage plot, echo=FALSE, message=FALSE, warning=FALSE}

peak <- loadNarrowPeaks( "U2OS_narrowpeaks", c("U2OSRNAPIIdmso1_callpeaknoabC", "U2OSRNAPIIetop_callpeaknoabC", "U2OSBRCA1nt_callpeaknoabC", "U2OSBRCA1etop_callpeaknoabC" ), peaks_suffix = "_peaks.narrowPeak", sub_dir = FALSE)

covplot <- covplot(peak, weightCol="score", chrs=c("chr17"), xlim=c(4.5e7, 5e7))
covplot + facet_grid(~factor (.id)) +
  theme(legend.position="none")

```

# Profile of ChIP peaks binding to TSS 

```{r Profile of ChIP peaks binding to TSS and gene body regions, echo=FALSE, message=FALSE, warning=FALSE}

promoter <- getPromoters(TxDb=txdb, upstream=3000, downstream=3000)

tagMatrix <- lapply(peak, getTagMatrix, windows=promoter)

# lapply(names(tagMatrix), function(x) plotAvgProf(tagMatrix[[x]], xlim=c(-3000, 3000),
#                                                  xlab="Genomic Region (5'->3')", ylab =
#                                                    "Read Count Frequency"))
# 
# # lapply(names(tagMatrix), function(x) tagHeatmap(tagMatrix[[x]], xlab="Genomic Region (5'-3')", title=names(tagMatrix[x])))
# 
```



```{r Transcriptional Start Site Comparison, echo=FALSE, message=FALSE, warning=FALSE}

RNAPIIcompareTSS <- plotAvgProf(tagMatrix[1:2], xlim=c(-3000,3000))  +
  scale_color_manual(values = c("blue", "red"))

BRCA1compareTSS <- plotAvgProf(tagMatrix[3:4], xlim=c(-3000,3000))  +
  scale_color_manual(values = c("blue", "red"))

RNAPIIcompareTSS 
BRCA1compareTSS

# plotAvgProf(tagMatrix, xlim=c(-3000, 3000), facet="row") + scale_y_continuous(limits = c(0, 0.002))

plotPeakProf2(peak, upstream = rel(0.2), downstream = rel(0.2),
              by = "gene", type = "body",
              TxDb = txdb, facet = "row", nbin = 800) + scale_y_continuous(limits = c(0, 0.02))

```

```{r ChIP peak annotation comparison, echo=FALSE, message=FALSE, warning=FALSE}

peakAnnoList <- lapply(peak, annotatePeak, TxDb=txdb,
                       tssRegion=c(-3000, 3000), annoDb="org.Hs.eg.db")


plotAnnoBar(peakAnnoList)


```

```{r Distribution of transcription factor-binding loci relative to TSS, echo=FALSE, message=FALSE, warning=FALSE}


plotDistToTSS(peakAnnoList)


```

# Enrichment Analysis

```{r Functional enrichment analysis, echo=FALSE, message=FALSE, warning=FALSE}

txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
peak <- loadNarrowPeaks( "U2OS_narrowpeaks", c("U2OSRNAPIIdmso1_callpeaknoabC", "U2OSRNAPIIetop_callpeaknoabC", "U2OSBRCA1nt_callpeaknoabC", "U2OSBRCA1etop_callpeaknoabC" ), peaks_suffix = "_peaks.narrowPeak", sub_dir = FALSE)

##Global

peakAnnoList <- lapply(peak, annotatePeak, TxDb=txdb,
                       tssRegion=c(-1000, 1000), flankDistance = 3000, level = "transcript", overlap='all')

peakannolistDF<- lapply(peakAnnoList, as.data.frame)

peakAnnoListgeneId<- lapply(peakannolistDF, pull, geneId)

enrichlist<- lapply(peakAnnoListgeneId, enrichPathway)
names(enrichlist)<-c(names(peakAnnoListgeneId))

##Promoter 

peakAnnoListPromoter <- lapply(names(peakannolistDF), function(x) filter(peakannolistDF[[x]], 
                                                                         grepl("Promoter", annotation)))
                               
peakAnnoListPromotergeneId<- lapply(peakAnnoListPromoter, pull, geneId)

enrichlistPromoter<- lapply(peakAnnoListPromotergeneId, enrichPathway)

names(enrichlistPromoter)<-c(names(peakannolistDF))

##Tables of associated pathways 

#Global

lapply(enrichlist, as.data.frame) %>%
  kbl(caption = htmltools::tags$caption("Global Reactome Pathway Enrichment", style = "color:black")) %>%
  kable_styling(bootstrap_options = "striped", full_width = T, html_font = "Cambria")

#Promoter

lapply(enrichlistPromoter, as.data.frame) %>%
  kbl(caption = htmltools::tags$caption("Promoter Reactome Pathway Enrichment", style = "color:black")) %>%
  kable_styling(bootstrap_options = "striped", full_width = T, html_font = "Cambria")

##Plotting


#BRCA1 global vs promoter only associated genes dmso

ggarrange(dotplot(enrichlist[["U2OSBRCA1nt_callpeaknoabC"]], showCategory=10), dotplot(enrichlistPromoter[["U2OSBRCA1nt_callpeaknoabC"]], showCategory=10), nrow = 2, ncol = 1, align = 'v')

#BRCA1 global vs promoter only associated genes etoposide

ggarrange(dotplot(enrichlist[["U2OSBRCA1etop_callpeaknoabC"]], showCategory=10), dotplot(enrichlistPromoter[["U2OSBRCA1etop_callpeaknoabC"]], showCategory=10), nrow = 2, ncol = 1, align = 'v')


#U2OSRNAPII global vs promoter only associated genes etoposide
ggarrange(dotplot(enrichlist[["U2OSRNAPIIetop_callpeaknoabC"]], showCategory=11), dotplot(enrichlistPromoter[["U2OSRNAPIIetop_callpeaknoabC"]], showCategory=11), nrow = 2, ncol = 1, align = 'v')

#U2OSRNAPII global vs promoter only associated genes dmso

ggarrange(dotplot(enrichlist[["U2OSRNAPIIdmso1_callpeaknoabC"]], showCategory=11), dotplot(enrichlistPromoter[["U2OSRNAPIIdmso1_callpeaknoabC"]], showCategory=11), nrow = 2, ncol = 1, align = 'v')





# lapply(enrichlistPromoter, as.data.frame)
# 
# dotplot(enrichlistPromoter[[4]])
# 
# 
# peakAnnoListPromoter<- lapply(peak, annotatePeak, TxDb=txdb,
#                        tssRegion=c(-3000, 3000), flankDistance = 3000, genomicAnnotationPriority="Promoter")
# 
# 
# 
# 
# ##Global
# 
# geneglobal<- lapply(peak, seq2gene, tssRegion = c(-1000, 1000), flankDistance=0, TxDb=txdb)
# 
# 
# enrichgeneglobal <- lapply(geneglobal, enrichPathway)
# 
# 
# enrichlistglobal<- lapply(names(peakAnnoList), function(x)
#   enrichPathway(as.data.frame(peakAnnoList[[x]])$geneId))
# 
# enrichlistpromoter<- lapply(names(peakAnnoListPromoter), function(x)
#   enrichPathway(as.data.frame(peakAnnoListPromoter[[x]])$geneId))
# 
# peakannolistDF<- lapply(peakAnnoList, as.data.frame)
# head(peakannolistDF)
# lapply(enrichlistglobal, as.data.frame)
# dotplot(enrichlistglobal[[4]]) + dotplot(enrichgeneglobal[[4]])
# 
# peakAnnoListPromoter<- lapply(peakAnnoList, function(x) as.data.frame(x))
# peakAnnoListPromoter<- lapply(names(peakAnnoList), function(x) filter(as.data.frame(peakAnnoList[[x]]), grepl("Promoter", annotation, ignore.case = TRUE)))
# genePromoter <- lapply(peakAnnoListPromoter, function(x) peakAnnoListPromoter[[x]]$geneId)
# enrichlistpromoter <- lapply(genePromoter, function(x) enrichPathway(genePromoter[[x]]))
# 
# RNAPIIDMSOenrichPromoter<-enrichPathway(peakAnnoListPromoter[[1]]$geneId)
# dotplot(RNAPIIDMSOenrichPromoter)
# RNAPIIetopenrichPromoter<-enrichPathway(peakAnnoListPromoter[[2]]$geneId)
# dotplot(RNAPIIetopenrichPromoter)
# 
# 
# filter(grepl("Promoter", annotation, ignore.case = TRUE))
#   
#   as.data.frame(peakAnnoRNAPIIetop) %>%
#   filter(str_detect(annotation, "Promoter"))
# 
# peakAnnoListPromoter <- lapply(names(peakannolistDF), function(x) filter(peakannolistDF[[x]], 
#                                                                          str_extract_all("<=1kb")))
#                                
# peakAnnoListPromotergeneId<- lapply(peakAnnoListPromoter, pull, geneId)
# 
# enrichlistPromoter<- lapply(peakAnnoListPromotergeneId, enrichPathway)
# 
# 
# ggarrange(dotplot(enrichlistPromoter[[1]]), dotplot(enrichlistPromoter[[2]]), dotplot(enrichlistPromoter[[3]]), dotplot(enrichlistPromoter[[4]]), ncol=1, nrow=4, align='v')
# 
# 
# enrichpathwaylist<- lapply(peakAnnoList, enrichPathway)
# enrichpathwaylistdf<-lapply(enrichpathwaylist, function(i) as.data.frame(i))
# enrichListdf <- mapply(cbind, enrichpathwaylistdf, "Variable"=names(enrichpathwaylist), SIMPLIFY=F)
# enrichListdf <- lapply(enrichListdf, FUN = function(X){X[c(10, 2, 1, 3:9)]})
# 
# enrichListDF
# 
# dotplotglobalRNAPIIdmso<-dotplot(enrichpathwaylist[["U2OSRNAPIIdmso1_callpeaknoabC"]], title="U2OSRNAPIIdmso Global Reactome Enrichment")+
#   xlab(NULL)
# dotplotglobalRNAPIIEtop<-dotplot(enrichpathwaylist[["U2OSRNAPIIetop_callpeaknoabC"]], title="U2OSRNAPIIEtop Global Reactome Enrichment")
# dotplotglobalBRCA1dmso<- dotplot(enrichpathwaylist[["U2OSBRCA1nt_callpeaknoabC"]], title="U2OSBRCA1dmso Global Reactome Enrichment")+
#   xlab(NULL)
# dotplotglobalBRCA1Etop<- dotplot(enrichpathwaylist[["U2OSBRCA1etop_callpeaknoabC"]], title="U2OSBRCA1Etop Global Reactome Enrichment")
# 
# ggarrange(dotplotglobalRNAPIIdmso, dotplotglobalRNAPIIEtop, ncol=1, nrow=2, align = 'v')
# ggarrange(dotglobalplotBRCA1dmso, dotploglobaltBRCA1Etop, ncol=1, nrow=2, align='v')
# 
# compReactomeDF<- lapply(compReactome, function(i) as.data.frame(i))
# compReactome <- compareCluster(geneglobal, fun="enrichPathway", pvalueCutoff= 0.05, pAdjustMethod="BH")
# dotplot(compReactome)
# 
# 
# # enrichlistglobal<- lapply(names(peakAnnoList), function(x) enrichPathway(as.data.frame(peakAnnoList[[x]])$geneId))
# # names(enrichlistglobal)<-c(names(peakAnnoList))
# # enrichlistglobalDF<-lapply(enrichlistglobal, function(i) as.data.frame(i))
# # enrichListglobalDF <- mapply(cbind, enrichListglobalDF, "Variable"=names(enrichlistglobal), SIMPLIFY=F)
# # enrichListglobalDF <- lapply(enrichListglobalDF, FUN = function(X){X[c(10, 2, 1, 3:9)]})
# 
# # enrichListglobalDF
# # 
# # peakAnnoListDF <- lapply(names(peakAnnoList), function(x) as.data.frame(peakAnnoList[[x]]))
# # 
# # peakAnnoListDF
# 
# ##Promoter only ; -1000, 1000 bp from TSS
# # peakAnnoPromoter<- lapply(names(peakAnnoList), function(x) as.data.frame(peakAnnoList[[x]]) %>%
# #                             filter(str_detect(annotation, "Promoter")))
# # enrichlistpromoter<-lapply(names(peakAnnoPromoter), function(x) enrichPathway(as.data.frame(peakAnnoPromoter[[x]])$geneId))
# # names(enrichlistpromoter) <- c("RNAPIIdmsoPromoterReactome", "RNAPIIetopPromoterReactome", "BRCA1dmsoPromoterReactome", "BRCA1etopPromoterReactome")
# # enrichlistpromoterDF<-lapply(enrichlistpromoter, function(i) as.data.frame(i))
# # enrichListpromoterDF <- mapply(cbind, enrichlistpromoterDF, "Variable"=names(enrichlistpromoter), SIMPLIFY=F)
# # enrichListpromoterDF
# # 
# # peakAnnoPromoterDF <- lapply(peakAnnoPromoter, function(i) as.data.frame(i))
# # peakAnnoPromoterDF <- mapply(cbind, peakAnnoPromoterDF, "Variable"=names(peakAnnoPromoter), SIMPLIFY=F)
# # 
# # 
# 
# geneshorter<- lapply(peak, seq2gene, tssRegion = c(-1000, 1000), flankDistance=0, TxDb=txdb)
# compReactomeshorter <- compareCluster(geneshorter, fun="enrichPathway", pvalueCutoff= 0.05, pAdjustMethod="BH")
# dotplot(compReactome)
# enrichpathwaylistshorter<- lapply(geneshorter, enrichPathway)
# enrichpathwaylistshorterdf<-lapply(enrichpathwaylistshorter, function(i) as.data.frame(i))
# enrichListshorterDFnewcol <- mapply(cbind, enrichpathwaylistshorterdf, "Variable"=names(enrichpathwaylistshorter), SIMPLIFY=F)
# enrichListDFnewcol <- lapply(enrichListshorterDFnewcol, FUN = function(X){X[c(10, 2, 1, 3:9)]})
# 
# enrichListDFnewcol
# 
# dotplotRNAPIIdmso<-dotplot(enrichpathwaylistshorter[["U2OSRNAPIIdmso1_callpeaknoabC"]], title="U2OSRNAPIIdmso Reactome Enrichment")+
#   xlab(NULL)
# dotplotRNAPIIEtop<-dotplot(enrichpathwaylistshorter[["U2OSRNAPIIetop_callpeaknoabC"]], title="U2OSRNAPIIEtop Reactome Enrichment")
# dotplotBRCA1dmso<- dotplot(enrichpathwaylistshorter[["U2OSBRCA1nt_callpeaknoabC"]], title="U2OSBRCA1dmso Reactome Enrichment")+
#   xlab(NULL)
# dotplotBRCA1Etop<- dotplot(enrichpathwaylistshorter[["U2OSBRCA1etop_callpeaknoabC"]], title="U2OSBRCA1Etop Reactome Enrichment")
# 
# ggarrange(dotplotRNAPIIdmso, dotplotRNAPIIEtop, ncol=1, nrow=2, align = 'v')
# ggarrange(dotplotBRCA1dmso, dotplotBRCA1Etop, ncol=1, nrow=2, align='v')
# 
# 
# 
# ##functional profile analysis 
# 
# peakAnnoListshorter <- lapply(peak, annotatePeak, TxDb=txdb,
#                        tssRegion=c(-1000, 1000), verbose=FALSE)
# 
# peakAnnoListDFPromoters <- lapply(peakAnnoListshorter, function(i) as.data.frame(i) %>%
#                            filter(str_detect(annotation, "Promoter")))
# 
# peakAnnoListDF <- lapply(peakAnnoList, function(i) as.data.frame(i))
# genesDF<- lapply(peakAnnoListDF, function(i) as.data.frame(i)$geneId)
# 
# 
# pathwaylist<- lapply(genesPromoter, enrichPathway)
# seq2genlist<- lapply(peak, seq2gene, tssRegion = c(-1000, 1000), flankDistance = 3000, TxDb=txdb)
# compReactome <- compareCluster(geneshorter, fun="enrichPathway", pvalueCutoff= 0.05, pAdjustMethod="BH")
# as.data.frame(compReactome)%>%
#   filter(str_detect(Description, "DNA Repair"))
#   
# compReactomeDF<- lapply(compReactome, function(i) as.data.frame(i))
# as.data.frame(compReactome@compareClusterResult)
# dotplot(compReactome, showCategory = 20)

```

```{r Overlap of peaks and annotated genes, echo=FALSE, message=FALSE, warning=FALSE}

genes <- lapply(peakAnnoListPromoter, function(i) as.data.frame(i)$geneId)
names(genes)<-c(names(peakannolistDF))
ggVennDiagram(c(genes[1], genes[3]), label = 'count')
ggVennDiagram(genes, label = 'count')

```


# Overlap of annotated genes

```{r Overlap of RNAPII dmso + BRCA1 dmso, echo=FALSE, message=FALSE, warning=FALSE}



ggarrange(ggVennDiagram(c(genes[1], genes[3]), label = "count"), ggVennDiagram(c(genes[2], genes[3]), label = "count"), ggVennDiagram(c(genes[1], genes[4]), label = "count"),
          ggVennDiagram(c(genes[2], genes[4]), label = "count"), nrow = 4, ncol = 1)

ggVennDiagram(genes, label = "count")

```

# Peak overlap enrichment analysis 

```{r Overlap of RNAPII dmso + BRCA1 dmso, echo=FALSE, message=FALSE, warning=FALSE}

overlap<-get.venn.partitions(genes[1:2])
overlaplist<- list(overlap$..values..)

names(overlaplist)<- names(overlap$..set..)

overlaplist <- split(overlap$..values.., seq(nrow(overlap)))
overlaplistnames<- split(overlap$..set.., seq(nrow(overlap)))

overlaplist[names(overlaplistnames)]<- overlaplist



```






